### PR TITLE
ApiExtractor now creates ApiDocumentation after all members have been created

### PR DESCRIPTION
**Pull request by dgaeta**
Refactored the creation of documention for ApiItems to occur after the abstract syntax tree has been created. This will be useful for implementing local API reference inheritdocs.

### CODE REVIEW

File: api-extractor/src/Extractor.ts [hunk 1]
```diff
@@ -95,7 +95,8 @@ export default class Extractor {
       throw new Error('Unable to load file: ' + options.entryPointFile);
     }
 
-    this.package = new ApiPackage(this, rootFile);
+    this.package = new ApiPackage(this, rootFile); // construct members
+    this.package.resolveReferences(); // creates ApiDocumentation
```
Comment (pgonzal) [inline api-extractor/src/Extractor.ts]: Do you need to update the Gulp task?  This is a breaking API change

File: api-extractor/src/definitions/ApiDocumentation.ts [hunk 2]
```diff
@@ -180,6 +181,13 @@ export default class ApiDocumentation {
   public readonly?: boolean;
 
   public docItemLoader: DocItemLoader;
+
+  /**
+   * We need the extractor to access the package that this ApiItem 
+   * belongs to in order to resolve references.
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiDocumentation.ts]: Can we obtain that via ApiItem.package.extractor?

File: api-extractor/src/definitions/ApiItem.ts [hunk 3]
```diff
@@ -52,6 +53,27 @@ export enum ApiItemKind {
 }
 
 /**
+ * The state of resolving the ApiItem's doc comment references.
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiItem.ts]: Maybe reference the resolveReferences() stage

File: api-extractor/src/definitions/ApiItem.ts [hunk 4]
```diff
@@ -52,6 +53,27 @@ export enum ApiItemKind {
...
+ */
+enum ResolveState {
+  /**
+   * The references of this ApiItem have not begun to be resolved.
+   */
+  Unresolved = 0,
+  /**
+   * The refernces of this ApiItem are in the process of being resolved.
+   * If we encounter this state again during resolution, a circular dependency is
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiItem.ts]: Maybe clarify "inside a recursive call to ApiItem.resolveReferences()"

File: api-extractor/src/definitions/ApiItem.ts [hunk 5]
```diff
@@ -237,6 +264,41 @@ abstract class ApiItem {
...
   }
+
+  /**
+   * This function assumes all references from this ApiItem have been resolved and we can now safely create 
+   * the documentation.
+   */
+  @virtual
+  protected onResolveReferences(): void {
+    this.documentation = new ApiDocumentation(this, this.extractor.docItemLoader, this.extractor, this.reportError);
+    // this.collectTypeReferences(this);
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiItem.ts]: ?

File: api-extractor/src/definitions/ApiItem.ts [hunk 6]
```diff
@@ -237,6 +264,41 @@ abstract class ApiItem {
...
+  }
+
+  /**
+   * Some ApiItems refer to the documentation of other ApiItems that are within the same package. To do that, 
+   * we have to make sure that the documentation is resolved for the reffered to ApiItem before we can 
+   * refer to it. This function makes sure we create the documentation for each ApiItem in the correct order.
+   * In the event that a circular dependency occurs, an error is reported. For example, if ApiItemOne has 
+   * an \@inheritdoc referencing ApiItemTwo, and ApiItemTwo has an \@inheritdoc refercing ApiItemOne then 
+   * we have a circular dependency and an error will be reported.
+   */
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiItem.ts]: Also clarify that resolveReferences() is a second stage that happens after the Analyzer calls the ApiItem constructors to build up the tree.

File: api-extractor/src/definitions/ApiItem.ts [hunk 7]
```diff
@@ -237,6 +264,41 @@ abstract class ApiItem {
...
+    switch (this._state) {
+      case ResolveState.Resolved:
+        return;
+      case ResolveState.Unresolved:
+        this._state = ResolveState.Resolving;
+        this.onResolveReferences();
+        this._state = ResolveState.Resolved;
+        return;
+      case ResolveState.Resolving:
+        this.reportError('@inheritdoc has a circular reference');
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiItem.ts]: \@inheritdoc isn't the only thing that can call resolveReferences(); this error could be misleading.

It also might be useful to create a couple errors and look at the line numbers and error message, to make sure it makes sense.

File: api-extractor/src/definitions/ApiItemContainer.ts [hunk 8]
```diff
@@ -25,6 +26,17 @@ abstract class ApiItemContainer extends ApiItem {
   protected addMemberItem(apiItem: ApiItem): void {
     this.memberItems.push(apiItem);
   }
+
+  /**
+   * Crawls the abstract syntax tree to resolve references.
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiItemContainer.ts]: For an @overrride function, you can use {@inheritdoc} :-P

File: api-extractor/src/definitions/ApiItemContainer.ts [hunk 9]
```diff
@@ -25,6 +26,17 @@ abstract class ApiItemContainer extends ApiItem {
...
+   */
+  @override
+  protected onResolveReferences(): void {
+    super.onResolveReferences();
+    this.memberItems.forEach(apiItem => {
+      apiItem.resolveReferences();  // <====  crawling tree
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiItemContainer.ts]: crawling tree?

File: api-extractor/src/definitions/ApiMethod.ts [hunk 10]
```diff
@@ -26,14 +26,12 @@ export default class ApiMethod extends ApiMember {
       this.params = [];
       for (const param of methodDeclaration.parameters) {
         const declarationSymbol: ts.Symbol = TypeScriptHelpers.tryGetSymbolForDeclaration(param);
-        const docComment: string = declarationSymbol && this.documentation && this.documentation.paramDocs ?
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiMethod.ts]: Could you run this on sp-core-library and make sure the output JSON doesn't have any diff from the last release?  That would help confirm that we don't need this code.

File: api-extractor/src/definitions/ApiPackage.ts [hunk 11]
```diff
@@ -74,8 +74,43 @@ export default class ApiPackage extends ApiItemContainer {
     }
   }
 
+  /**
+   * Find a member in this package by name and return it if found.
+   * 
+   * @param memberName - the name of the member ApiItem
+   */
+  protected getMemberItem(memberName: string): ApiItem {
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiPackage.ts]: Is this used in this PR?

File: api-extractor/src/definitions/ApiPackage.ts [hunk 12]
```diff
@@ -74,8 +74,43 @@ export default class ApiPackage extends ApiItemContainer {
...
+   * do so, we need to resolve any references this member refers to.  
+   *
+   * Example, if an ApiItem 'A' references the documentation of ApiItem 'B' but 
+   * 'B' does not have documentation created yet, we need to create the documentation
+   * for 'B' first. So we would call this function and provided the name of ApiItem 'B'
+   * to ensure its documentation is created first.
+   * 
+   * @param memberName - the name of the member we want to ensure has documentation created.
+   */
+  public resolveReferencesOnItem(memberName: string): ApiItem {
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiPackage.ts]: This looks weird.  Maybe we should bump it to the next PR if nothing is using it.

File: api-extractor/src/definitions/ApiItem.ts [hunk 13]
```diff
@@ -2,6 +2,7 @@
 /* tslint:disable:no-constant-condition */
 
 import * as ts from 'typescript';
+import { virtual } from '@microsoft/decorators/lib/virtual';
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiItem.ts]: Don't use path-based imports

File: api-extractor/src/Extractor.ts (hunk 1 above)
Comment (dgaeta) [inline api-extractor/src/Extractor.ts]: We don't this change is contained within the Extractor.analyze() function so it is hidden to the Gulp task.

File: api-extractor/src/definitions/ApiDocumentation.ts (hunk 2 above)
Comment (dgaeta) [inline api-extractor/src/definitions/ApiDocumentation.ts]: ApiItem doesn't have a package property.

**Pull request merged**